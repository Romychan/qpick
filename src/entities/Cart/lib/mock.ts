// eslint-disable-next-line @conarti/feature-sliced/layers-slices
import { Product } from '~/entities/Product/@x';

import { CartItem } from '../model/types';

/** Mock product for the cart */
export const MOCK_CART_PRODUCT: Product = {
  id: 1,
  title: 'Product',
  rate: 4.2,
  image: '/images/catalog/headphone-1.png',
  price: 1200,
  oldPrice: 1300,
  description: 'Test',
  category: 'Наушники',
};

/** List of mock cart items */
export const MOCK_CART: CartItem[] = [
  {
    product: MOCK_CART_PRODUCT,
    quantity: 2,
  },
];
