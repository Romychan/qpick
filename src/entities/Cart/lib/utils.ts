// eslint-disable-next-line @conarti/feature-sliced/layers-slices
import { Product } from '~/entities/Product/@x';

import { CartItem } from '../model/types';

/**
 * Add a product to the cart
 *
 * @param cartItems The current state of the cart
 * @param product Product to add to cart
 * @returns Updated state of the product cart
 */
export const addToCart = (cartItems: CartItem[], product: Product) => {
  const productInCart = cartItems.find(
    (item) => item.product.id === product.id,
  );

  if (!productInCart) {
    return [...cartItems, { product, quantity: 1 }];
  } else {
    return cartItems.map((item) =>
      productInCart.product.id === item.product.id
        ? { ...productInCart, quantity: item.quantity + 1 }
        : item,
    );
  }
};

/**
 * Delete a product from the cart
 *
 * @param cartItems The current state of the cart
 * @param id The ID of the product to be deleted from the cart
 * @returns Updated state of the product cart
 */
export const deleteFromCart = (cartItems: CartItem[], id: number) => {
  return cartItems.filter((item) => item.product.id !== id);
};

/**
 * Increase the number of product in the cart
 *
 * @param cartItems The current state of the cart
 * @param id Product ID to increase the quantity
 * @returns Updated state of the product cart
 */
export const increaseProductQuantity = (cartItems: CartItem[], id: number) => {
  const productInCart = cartItems.find((item) => item.product.id === id);

  if (productInCart) {
    return cartItems.map((item) =>
      productInCart.product.id === item.product.id
        ? { ...productInCart, quantity: item.quantity + 1 }
        : item,
    );
  }

  return cartItems;
};

/**
 * Decrease the number of product in the cart
 *
 * @param cartItems The current state of the cart
 * @param id Product ID to decrease the quantity
 * @returns Updated state of the product cart
 */
export const decreaseProductQuantity = (cartItems: CartItem[], id: number) => {
  const productInCart = cartItems.find((item) => item.product.id === id);

  if (productInCart) {
    return cartItems.map((item) =>
      productInCart.product.id === item.product.id
        ? { ...productInCart, quantity: item.quantity - 1 }
        : item,
    );
  }

  return cartItems;
};
