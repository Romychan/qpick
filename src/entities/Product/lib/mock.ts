import { Product } from '../model/types';

/** List of mock products */
export const MOCK_PRODUCTS: Product[] = [
  {
    id: 1,
    title: 'Apple BYZ S852I',
    rate: 4.7,
    image: '/images/catalog/headphone-1.png',
    price: 2927,
    oldPrice: 3527,
    description:
      'Lorem Ipsum is simply dummy text of the printing and typesetting industry.',
    category: 'Наушники',
  },
  {
    id: 2,
    title: 'Apple AirPods Pro 3',
    rate: 4.5,
    image: '/images/catalog/headphone-2.png',
    price: 200,
    oldPrice: null,
    description:
      "Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.",
    category: 'Наушники',
  },
  {
    id: 3,
    title: 'Apple EarPods',
    rate: 4.5,
    image: '/images/catalog/headphone-3.png',
    price: 2327,
    oldPrice: null,
    description:
      'Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.',
    category: 'Наушники',
  },
  {
    id: 4,
    title: 'Apple BYZ S123I',
    rate: 4.7,
    image: '/images/catalog/headphone-1.png',
    price: 1500,
    oldPrice: 1950,
    description:
      'Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.',
    category: 'Наушники',
  },
  {
    id: 5,
    title: 'Apple AirPods Pro 2',
    rate: 4.5,
    image: '/images/catalog/headphone-2.png',
    price: 200,
    oldPrice: null,
    description:
      'Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.',
    category: 'Наушники',
  },
  {
    id: 6,
    title: 'Apple EarPods 2',
    rate: 3.8,
    image: '/images/catalog/headphone-3.png',
    price: 1200,
    oldPrice: 2000,
    description:
      'Aenean maximus, nisi eget mollis dictum, leo sapien sollicitudin ipsum, quis euismod ex erat ut tellus.',
    category: 'Наушники',
  },
  {
    id: 7,
    title: 'Apple AirPods',
    rate: 4.7,
    image: '/images/catalog/headphone-4.png',
    price: 9527,
    oldPrice: null,
    description:
      'Aenean maximus, nisi eget mollis dictum, leo sapien sollicitudin ipsum, quis euismod ex erat ut tellus.',
    category: 'Беспроводные наушники',
  },
  {
    id: 8,
    title: 'GERLAX GH-04',
    rate: 4.7,
    image: '/images/catalog/headphone-5.png',
    price: 6527,
    oldPrice: null,
    description:
      'Sed quis sem vitae justo feugiat tincidunt. Maecenas ut leo nisl. Praesent finibus ultricies ex et tristique.',
    category: 'Беспроводные наушники',
  },
  {
    id: 9,
    title: 'BOROFONE BO4',
    rate: 4.7,
    image: '/images/catalog/headphone-6.png',
    price: 7527,
    oldPrice: null,
    description:
      "Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.",
    category: 'Беспроводные наушники',
  },
];
