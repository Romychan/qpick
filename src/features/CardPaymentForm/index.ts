export { CardPaymentForm } from './ui/CardPaymentForm';
