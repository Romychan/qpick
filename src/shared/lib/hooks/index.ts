export { useEventListener } from './useEventListener';
export { useIsMounted } from './useIsMounted';
