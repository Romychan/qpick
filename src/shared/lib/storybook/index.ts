export { withReactHookFormDecorator } from './decorators/withReactHookFormDecorator';
export { withHeightDecorator } from './decorators/withHeightDecorator';
export { withGapDecorator } from './decorators/withGapDecorator';
export { withWidthDecorator } from './decorators/withWidthDecorator';
export { withRouterDecorator } from './decorators/withRouterDecorator';
