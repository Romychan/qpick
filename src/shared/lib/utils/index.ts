export { formatPrice, formatCardNumber, formatExpiryDate } from './formatters';
export { cl } from './className';
