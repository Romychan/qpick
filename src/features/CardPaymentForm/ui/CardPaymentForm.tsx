import { Controller, useFormContext } from 'react-hook-form';

import { Input } from '~/shared/ui/Input';
import { FormControl } from '~/shared/ui/FormControl';
import { formatCardNumber, formatExpiryDate } from '~/shared/lib/utils';

import styles from './CardPaymentForm.module.scss';

/** This is a component of the form for filling out information about a bank card */
export const CardPaymentForm = () => {
  const { control } = useFormContext();

  return (
    <div className={styles.container} data-testid="card-payment">
      <Controller
        name="payment.card"
        control={control}
        render={({
          field: { name, value, onChange },
          fieldState: { error },
        }) => (
          <FormControl
            label="Номер карты"
            labelId={name}
            error={error?.message}
            required
          >
            <Input
              id={name}
              name={name}
              placeholder="1234 1234 1234 1234"
              value={formatCardNumber(value)}
              onChange={onChange}
              error={!!error}
              minLength={4}
              maxLength={19}
            />
          </FormControl>
        )}
      />

      <div className={styles.row}>
        <Controller
          name="payment.expiryDate"
          control={control}
          render={({
            field: { name, value, onChange },
            fieldState: { error },
          }) => (
            <FormControl
              label="Срок действия"
              labelId={name}
              error={error?.message}
              required
            >
              <Input
                id={name}
                name={name}
                placeholder="ММ/ГГ"
                value={formatExpiryDate(value)}
                onChange={onChange}
                error={!!error}
                minLength={5}
                maxLength={5}
              />
            </FormControl>
          )}
        />

        <Controller
          name="payment.cvv"
          control={control}
          defaultValue=""
          render={({
            field: { name, value, onChange },
            fieldState: { error },
          }) => (
            <FormControl
              label="CVC/CVV"
              labelId={name}
              error={error?.message}
              required
            >
              <Input
                id={name}
                name={name}
                type="password"
                placeholder="***"
                value={value}
                onChange={onChange}
                error={!!error}
                minLength={3}
                maxLength={3}
              />
            </FormControl>
          )}
        />
      </div>
    </div>
  );
};
