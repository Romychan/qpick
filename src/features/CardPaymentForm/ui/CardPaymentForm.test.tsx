import { RenderResult, screen } from '@testing-library/react';

import { renderWithReactHookForm } from '~/shared/lib/tests';

import { CardPaymentForm } from './CardPaymentForm';

describe('CardPaymentForm', () => {
  let container: RenderResult;

  beforeEach(() => {
    container = renderWithReactHookForm(<CardPaymentForm />);
  });

  it('should render correctly', () => {
    expect(screen.getAllByTestId('input-field')).toHaveLength(3);
  });

  it('should match snapshot', () => {
    expect(container).toMatchSnapshot();
  });
});
