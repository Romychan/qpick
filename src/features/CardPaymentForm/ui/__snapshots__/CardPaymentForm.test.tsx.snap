// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CardPaymentForm > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_container_7becc1"
        data-testid="card-payment"
      >
        <div
          class="_container_77a9ed _primary_77a9ed _top_77a9ed"
        >
          <label
            class="_label_77a9ed"
            for="payment.card"
          >
            Номер карты
            <span
              class="_required_77a9ed"
            >
              *
            </span>
          </label>
          <div
            class="_content_77a9ed"
          >
            <input
              autocomplete="off"
              class="_input_95237c _bordered_95237c"
              data-testid="input-field"
              id="payment.card"
              maxlength="19"
              minlength="4"
              name="payment.card"
              placeholder="1234 1234 1234 1234"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="_row_7becc1"
        >
          <div
            class="_container_77a9ed _primary_77a9ed _top_77a9ed"
          >
            <label
              class="_label_77a9ed"
              for="payment.expiryDate"
            >
              Срок действия
              <span
                class="_required_77a9ed"
              >
                *
              </span>
            </label>
            <div
              class="_content_77a9ed"
            >
              <input
                autocomplete="off"
                class="_input_95237c _bordered_95237c"
                data-testid="input-field"
                id="payment.expiryDate"
                maxlength="5"
                minlength="5"
                name="payment.expiryDate"
                placeholder="ММ/ГГ"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="_container_77a9ed _primary_77a9ed _top_77a9ed"
          >
            <label
              class="_label_77a9ed"
              for="payment.cvv"
            >
              CVC/CVV
              <span
                class="_required_77a9ed"
              >
                *
              </span>
            </label>
            <div
              class="_content_77a9ed"
            >
              <input
                autocomplete="off"
                class="_input_95237c _bordered_95237c"
                data-testid="input-field"
                id="payment.cvv"
                maxlength="3"
                minlength="3"
                name="payment.cvv"
                placeholder="***"
                type="password"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_container_7becc1"
      data-testid="card-payment"
    >
      <div
        class="_container_77a9ed _primary_77a9ed _top_77a9ed"
      >
        <label
          class="_label_77a9ed"
          for="payment.card"
        >
          Номер карты
          <span
            class="_required_77a9ed"
          >
            *
          </span>
        </label>
        <div
          class="_content_77a9ed"
        >
          <input
            autocomplete="off"
            class="_input_95237c _bordered_95237c"
            data-testid="input-field"
            id="payment.card"
            maxlength="19"
            minlength="4"
            name="payment.card"
            placeholder="1234 1234 1234 1234"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="_row_7becc1"
      >
        <div
          class="_container_77a9ed _primary_77a9ed _top_77a9ed"
        >
          <label
            class="_label_77a9ed"
            for="payment.expiryDate"
          >
            Срок действия
            <span
              class="_required_77a9ed"
            >
              *
            </span>
          </label>
          <div
            class="_content_77a9ed"
          >
            <input
              autocomplete="off"
              class="_input_95237c _bordered_95237c"
              data-testid="input-field"
              id="payment.expiryDate"
              maxlength="5"
              minlength="5"
              name="payment.expiryDate"
              placeholder="ММ/ГГ"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="_container_77a9ed _primary_77a9ed _top_77a9ed"
        >
          <label
            class="_label_77a9ed"
            for="payment.cvv"
          >
            CVC/CVV
            <span
              class="_required_77a9ed"
            >
              *
            </span>
          </label>
          <div
            class="_content_77a9ed"
          >
            <input
              autocomplete="off"
              class="_input_95237c _bordered_95237c"
              data-testid="input-field"
              id="payment.cvv"
              maxlength="3"
              minlength="3"
              name="payment.cvv"
              placeholder="***"
              type="password"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
